### PR TITLE
[Feat] Add API support for core features

### LINKED ISSUE #54: feat: api - improve endpoint definitions
**Is this a bug report or feature request?** (choose one): feat

**Python Version** (output of `python --version`): 2.7

**Development or Deployment Environment?**:

**Release Tag or Master**:

**Expected Behavior**:

**What Actually Happened**:

**How to Reproduce the Issue** (as minimally as possible):

**Any Additional Comments**:

improve the api

### PR DESCRIPTION
**Pull request by drewwalters96**
- Refactor API structure
- Add API support for existing CLI flags
- Add Keystone token and RBAC

**What is the purpose of this pull request?**: To enhance the Armada API and introduce features already present in the CLI.

**What issue does this pull request address?**: Fixes #54 

**Notes for reviewers to consider**: 

**Specific reviewers for pull request**: @gardlt @theyer @sh8121att

### CODE REVIEW

File: armada/api/Armada.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# Copyright 2017 The Armada Authors.
```
Comment (gardlt) [inline armada/api/Armada.py]: armada.py not Armada.py

File: armada/api/middleware.py [hunk 2]
```diff
@@ -11,17 +11,43 @@
...
-        token = req.get_header('X-Auth-Token')
-        self.validate_token(token)
-
-    def validate_token(self, token):
-        if token is None:
-            raise falcon.HTTPUnauthorized('X-Auth token invalid.')
-        if token != 'armada':
-            raise falcon.HTTPUnauthorized('X-Auth token invalid.')
+        token = req.get_header('X-Auth-Token', required=True)
+        return self._client.tokens.validate(token=token)
```
Comment (sh8121att) [inline armada/api/middleware.py]: I don't think the return value of the middleware process_* methods have any impact. This middleware should use the Keystone token to get the authenticated username and roles and add them to a context object. The downstream handler for the request can then use the context to authorize the roles for the actions being requested.

File: armada/api/middleware.py [hunk 3]
```diff
@@ -40,49 +38,67 @@
...
-    def _password_auth(self):
-	auth = v3.Password(auth_url="http://192.168.148.91:35357/v3",
+	def process_request(self, req, resp):
+		token = req.get_header('X-Auth-Token')
+
+        # Validate token and get user session
+        #self._session = self._get_token_session(token)
+
+		# TODO: Remove method call (testing purposes)
+		self._session = self._password_auth()
```
Comment (sh8121att) [inline armada/api/middleware.py]: Note that a AuthMiddleware instance will be reused for all requests, so using `self._session` to validate a request token is probably unwise. Consider just using a variable local to this method and it injecting it into downstream methods like `_get_roles()`

File: armada/api/middleware.py [hunk 4]
```diff
@@ -40,49 +38,67 @@
...
+		# Get roles associated with user
+		try:
+			resp = self._session.get(request_url)
+		except:
+			raise falcon.HTTPUnauthorized('Authentication required',
+	                              ('Authentication token is invalid.'))
+								  
+		json_resp = resp.json()['role_assignments']
+
+		roles = []
```
Comment (sh8121att) [inline armada/api/middleware.py]: This can also use a list comprehension:
`roles = [r['role']['id'].encode('utf-8') for r in json_resp]`

Also consider adding this to the try/except block to catch any typing errors.

File: armada/api/middleware.py [hunk 5]
```diff
@@ -40,49 +38,67 @@
...
+		for role in json_resp:
+			roles.append(role['role']['id'].encode('utf-8'))
```
Comment (sh8121att) [inline armada/api/middleware.py]: I'm not sure if we will want the `id` or the role name. Our policy.yaml file will likely have something like
`'armada:apply_chart' : 'role:deployer'` stating to apply a chart in Armada you need the role called 'deployer'.

File: armada/api/middleware.py (hunk 3 above)
Comment (drewwalters96) [inline armada/api/middleware.py]: That makes sense. I will update this.

File: armada/api/middleware.py (hunk 4 above)
Comment (drewwalters96) [inline armada/api/middleware.py]: Thanks, I will also make this change.

File: armada/api/middleware.py (hunk 5 above)
Comment (drewwalters96) [inline armada/api/middleware.py]: This makes sense to me. @gardlt would you be opposed to doing it this way?

File: armada/conf/default.py [hunk 6]
```diff
@@ -17,6 +17,17 @@
...
 default_options = [
+
+    cfg.ListOpt(
+        'armada_apply_roles',
+        default=['admin'],
+        help='IDs of approved API access roles.'),
+
+    cfg.StrOpt(
+        'auth_url',
+        default='http://192.168.148.91/v3',
```
Comment (gardlt) [inline armada/conf/default.py]: prob make this more generic. (ie. keystone-api )

File: armada/conf/default.py [hunk 7]
```diff
@@ -105,6 +116,21 @@
         help='Defines the format string for \
         %(user_identity)s that is used in logging_context_format_string.'),
 
+    cfg.StrOpt(
+        'project_domain_name',
```
Comment (gardlt) [inline armada/conf/default.py]: We should add an option for the CIDR:

File: examples/openstack-helm.yaml [hunk 8]
```diff
@@ -113,20 +113,3 @@ armada:
             reference: master
           dependencies:
           - *helm-toolkit
-      - chart: &horizon
-          name: horizon
```
Comment (gardlt) [inline examples/openstack-helm.yaml]: why is this removed.

File: armada/conf/default.py (hunk 6 above)
Comment (gardlt) [inline armada/conf/default.py]: make this more generic (ie. `keystone-api` or `0.0.0.0` ) not dev api.

File: armada/conf/default.py [hunk 9]
```diff
@@ -105,6 +116,21 @@
...
+        default='default',
+        help='The Keystone project domain name used for authentication.'),
+
+    cfg.StrOpt(
+        'project_name',
+        default='admin',
+        help='The Keystone project name used for authentication.'),
+
+    cfg.PortOpt(
```
Comment (gardlt) [inline armada/conf/default.py]: Like that we are managing the port via the config file can we also do it for the CIDR (ie. 0.0.0.0 )

File: armada/handlers/__init__.py [hunk 10]
```diff
@@ -11,3 +11,13 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
```
Comment (gardlt) [inline armada/handlers/__init__.py]: do we want to remove this from the version file?

**Comment by gardlt** [comment]
This is a good start just some nit pick.

File: armada/conf/default.py [hunk 11]
```diff
@@ -17,6 +17,17 @@
 from oslo_config import cfg
 
 default_options = [
+
+    cfg.ListOpt(
+        'armada_apply_roles',
+        default=['admin'],
+        help='IDs of approved API access roles.'),
+
+    cfg.StrOpt(
```
Comment (intlabs) [inline armada/conf/default.py]: rather than manually creating the keystone options manually, can we leverage oslo.config and import these options, thus exposing every param and reducing ongoing tech-debt.

Comment (drewwalters96) [inline armada/conf/default.py]: That makes a lot of sense to me, although I included these options because we use these to generate a default config file. Do you think it would be beneficial to have these options included in a default config file that is generated by Armada?

Comment (drewwalters96) [inline armada/conf/default.py]: After discussing with @gardlt, we will create an issue for this suggestion and iterate on it
